### ISSUE TITLE
[v2.0-c11] move core implementation to C11

### ISSUE THREAD

**Issue by blueloveTH** [pocketpy/pocketpy#265]
## Description
1. Redesign the core architecture and move them to C11
2. Let C++ classes wraps new C11 APIs

## Headers
- [ ] include/pocketpy/common/memorypool.hpp: 26
- [ ] include/pocketpy/common/version.h: 7
- [ ] include/pocketpy/common/export.h: 40
- [ ] include/pocketpy/common/str.hpp: 244
- [ ] include/pocketpy/common/config.h: 67
- [ ] include/pocketpy/common/vector.hpp: 458
- [ ] include/pocketpy/common/traits.hpp: 40
- [ ] include/pocketpy/common/any.hpp: 113
- [ ] include/pocketpy/common/gil.hpp: 20
- [ ] include/pocketpy/common/utils.hpp: 36
- [ ] include/pocketpy/common/types.hpp: 37
- [ ] include/pocketpy/common/_generated.hpp: 21
- [ ] include/pocketpy/objects/codeobject.hpp: 180
- [ ] include/pocketpy/objects/sourcedata.hpp: 29
- [ ] include/pocketpy/objects/error.hpp: 89
- [ ] include/pocketpy/objects/object.hpp: 39
- [ ] include/pocketpy/objects/dict.hpp: 66
- [ ] include/pocketpy/objects/base.hpp: 110
- [ ] include/pocketpy/objects/namedict.hpp: 64
- [ ] include/pocketpy/objects/tuplelist.hpp: 77
- [ ] include/pocketpy/objects/builtins.hpp: 147
- [ ] include/pocketpy/objects/stackmemory.hpp: 18

## Sources
- [ ] src/common/any.cpp: 26
- [ ] src/common/memorypool.cpp: 297
- [ ] src/common/str.cpp: 578
- [ ] src/objects/tuplelist.cpp: 56
- [ ] src/objects/builtins.cpp: 6
- [ ] src/objects/error.cpp: 21
- [ ] src/objects/object.cpp: 5
- [ ] src/objects/dict.cpp: 180
- [ ] src/objects/namedict.cpp: 163
- [ ] src/objects/sourcedata.cpp: 68
- [ ] src/objects/codeobject.cpp: 9

**Comment by blueloveTH** [comment]
## Coding Style Tips

+ By default, the new C11 functions follows this name: `pkpy_{class}__{function}`
+ Use `T* self` as the first parameter if the function is a previous C++ method
+ Do not change `PyVar` and `PyObject`
+ Do not change Enums
+ Use comment `T=` to annotate generic array and vector, e.g. `c11_vector/* T=c11_string */`
+ Use a single line `//` comment for function if it is not clear

**Comment by blueloveTH** [comment]
TODO:
1. rename `::PyVar` to `pkpy_Value`
2. rename `::PyObject`

**Comment by nathanpage-credo** [comment]
It looks like you have dropped support for multiple VMs in v2 and instead use `pk_current_vm`. Is that a feature that will come back?

**Comment by blueloveTH** [comment]
> It looks like you have dropped support for multiple VMs in v2 and instead use `pk_current_vm`. Is that a feature that will come back?

Paritially. v2 still supports multi-vm. You can keep multi-vm states but only one of them can be active. And all operations are based on the current active vm. We will introduce a API like `py_switch_vm` which allows you to switch between VMs.

**Comment by pinkiedreams** [comment]
> > It looks like you have dropped support for multiple VMs in v2 and instead use `pk_current_vm`. Is that a feature that will come back?
> 
> Paritially. v2 still supports multi-vm. You can keep multi-vm states but only one of them can be active. And all operations are based on the current active vm. We will introduce a API like `py_switch_vm` which allows you to switch between VMs.

It is much more convenient (and less memory-consuming) to pass VM* to function arguments, then the language can be fully considered embedded and interact with callback events

**Comment by OcJlik** [comment]
> > > It looks like you have dropped support for multiple VMs in v2 and instead use `pk_current_vm`. Is that a feature that will come back?
> > 
> > 
> > Paritially. v2 still supports multi-vm. You can keep multi-vm states but only one of them can be active. And all operations are based on the current active vm. We will introduce a API like `py_switch_vm` which allows you to switch between VMs.
> 
> It is much more convenient (and less memory-consuming) to pass VM* to function arguments, then the language can be fully considered embedded and interact with callback events

seriously? it is surprising that no one has yet thought of making a normal embedded python
this is the first good library but with this huge minus it is simply not needed

**Comment by ZeroLog** [comment]
> > > It looks like you have dropped support for multiple VMs in v2 and instead use `pk_current_vm`. Is that a feature that will come back?
> > 
> > 
> > Paritially. v2 still supports multi-vm. You can keep multi-vm states but only one of them can be active. And all operations are based on the current active vm. We will introduce a API like `py_switch_vm` which allows you to switch between VMs.
> 
> It is much more convenient (and less memory-consuming) to pass VM* to function arguments, then the language can be fully considered embedded and interact with callback events

I liked the idea of ​​the guy from Discord: load several .so and match exports with the control object. Unfortunately, this requires copying the .so file and greatly increases RAM consumption. It is much easier to add this by default to the library.

**Comment by Dunekeru** [comment]
> > > > It looks like you have dropped support for multiple VMs in v2 and instead use `pk_current_vm`. Is that a feature that will come back?
> > > 
> > > 
> > > Paritially. v2 still supports multi-vm. You can keep multi-vm states but only one of them can be active. And all operations are based on the current active vm. We will introduce a API like `py_switch_vm` which allows you to switch between VMs.
> > 
> > 
> > It is much more convenient (and less memory-consuming) to pass VM* to function arguments, then the language can be fully considered embedded and interact with callback events
> 
> I liked the idea of ​​the guy from Discord: load several .so and match exports with the control object. Unfortunately, this requires copying the .so file and greatly increases RAM consumption. It is much easier to add this by default to the library.

You can load .dll .so into memory. But yes, the memory consumption will be inadequate, and it is unlikely that such a standard will be adopted by a decent company.

**Comment by Achikaps** [comment]
> > > It looks like you have dropped support for multiple VMs in v2 and instead use `pk_current_vm`. Is that a feature that will come back?
> > 
> > 
> > Paritially. v2 still supports multi-vm. You can keep multi-vm states but only one of them can be active. And all operations are based on the current active vm. We will introduce a API like `py_switch_vm` which allows you to switch between VMs.
> 
> It is much more convenient (and less memory-consuming) to pass VM* to function arguments, then the language can be fully considered embedded and interact with callback events

I totally agree, the library is great in every way, except for this, we were given requirements at the game development studio and I really want this to be added, otherwise we'll have to add Lua, which I hate

**Comment by emp0ry** [comment]
> It is much more convenient (and less memory-consuming) to pass VM* to function arguments, then the language can be fully considered embedded and interact with callback events

Why wasn't this originally @blueloveTH? This is a normal implementation of any embedded language. At least answer what you want, otherwise we don't understand why the language is compared to Lua.
